### DISCUSSION TITLE
*GASP* A debuggable beartype?

### DISCUSSION THREAD

**Discussion by posita** [beartype/beartype#84] [Ideas]
I noticed that beartype-generated code is not visible to a debugger. This is unsurprising, but still not very convenient for explorations such as #67. Consider:

``` python
from beartype import beartype

@beartype
def foo(i: int) -> str:
    breakpoint()
    return str(i)

print(foo(0))
```

``` sh
% python test_case.py
> /…/test_case.py(6)foo()
-> return str(i)
(Pdb) l
  1     from beartype import beartype
  2
  3     @beartype
  4     def foo(i: int) -> str:
  5         breakpoint()
  6  ->     return str(i)
  7
  8     print(foo(0))
[EOF]
(Pdb) u
> <string>(29)foo()
(Pdb) l
[EOF]
(Pdb) u
> /…/test_case.py(8)<module>()
-> print(foo(0))
(Pdb) l
  3     @beartype
  4     def foo(i: int) -> str:
  5         breakpoint()
  6         return str(i)
  7
  8  -> print(foo(0))
[EOF]
(Pdb) c
0
```

That `<string>(29)foo()` is where sweet, golden honey lives, but it's _hidden_.

I've been casually thinking about https://github.com/beartype/beartype/commit/8fd38b36b13f0e653785166dab3edc2892a03d6d#commitcomment-59708937 and spotted a potential synergy with tricking the debugger into thinking it had beartype-generated source. I started looking into what that would take to remedy, and quickly found that it wasn't possible (or at least not obvious) using standard tooling. That led to me filing https://bugs.python.org/issue46133, which I expect will never be fixed.

That sat there for a good month or so. You'll never guess who _just_ came to my aid, but good ol' @TeamSpen210 ([yup](https://github.com/beartype/beartype/issues/66), [the](https://github.com/beartype/beartype/issues/62#issuecomment-952396797) [very](https://github.com/beartype/beartype/issues/54#issuecomment-939389619) [same](https://github.com/beartype/beartype/issues/60#issuecomment-950101704)) with [this gem](https://bugs.python.org/msg410640)!

> There is a solution to this: you can modify the linecache module's cache to add lines under a fake filename, which is the approach attrs takes here: https://github.com/python-attrs/attrs/blob/9727008fd1e40bc55cdc6aee71e0f61553f33127/src/attr/_make.py#L347
> However, there are several downsides.
>
> - That dict isn't documented or appears in __all__, so it's arguably private.
> - The C implementation of traceback printing opens files directly, so this doesn't work there.
> - You have to invent a unique filename, and manually remove lines from the cache if the function is later deleted.
>
> This does affect both namedtuple and dataclasses, though it's probably not too important given how straightforward the generated code they produce is.

Now, as Spencer points out, that ain't no public interface and it probably ain't no path for the faint of heart, neither. I am wondering if there is any appetite for me putting together a PR that might bring that `attrs` ~hack~ _technique_ to beartype. If it turns out to be expensive, it could always be gated (default closed) at runtime, as you mention in https://github.com/beartype/beartype/commit/8fd38b36b13f0e653785166dab3edc2892a03d6d#commitcomment-60594974. I'm not promising anything, but I wanted to throw it out there for some tire-kicking before I started messing around with it in earnest.

Thoughts?

**Comment by TeamSpen210** [comment]
The [`linecache`](https://github.com/python/cpython/blob/main/Lib/linecache.py) module is not too complex - all you need to do is invent a unique filename for every generated function, then add a `(size, mtime, line_list, fullname)` tuple to the dict. I'd suggest base64 of the original func's ID. Though it probably wouldn't be very common, we could also use `weakref.finalize()` to call `linecache.cache.pop(filename, None)` so the strings don't stay alive. 

It's not really expensive, but it'd cause a `splitlines()`-ed version of the code to be kept in memory, which might be a concern.

**Comment by leycec** [comment]
*...heh.* Brilliant AI-enhanced minds compute alike. :robot: 

As a cabin-dwelling neckbeard, I've actually known about [this exact issue and its exact solution for over a year](https://github.com/beartype/beartype/blob/609abc9e8a2b563c9848e2ef77ea2726aade4021/beartype/_decor/main.py#L302). As a cabin-dwelling neckbeard, I failed to do anything about it. Apparently squandering everyone's precious time like so much filthy sand through my sclerotic fingers was last year's New Year's Resolution. `</sigh>`

**This must happen.** The only way things happen is with public ~~beatings~~ shaming, so I've opened [a new feature request helpfully shaming me into finally doing this](https://github.com/beartype/beartype/issues/85).

<sup>"Why!? Why I never do the good things?"</sup>

> Though it probably wouldn't be very common, we could also use `weakref.finalize()` to call `linecache.cache.pop(filename, None)` so the strings don't stay alive. It's not really expensive, but it'd cause a `splitlines()`-ed version of the code to be kept in memory, which might be a concern.

This... this is good. I *never* considered that space consumption might be a concern here – but Spence's concerns have now become my concerns. Anxiety must be contagious, because I'm breathing heavily into a paper bag over here.

I'd be unconcerned if `linecache` was just intelligently caching (possibly interned) string literals. But `linecache` is naively caching mutable lists of strings produced by `str.splitlines()`, so space consumption suddenly becomes a fickle thing.

And the many-headed hydra named Igonna Break Yercode just grew another head.

**Comment by posita** [comment]
I love everything except the shame bit. Ain't no shame in this game.

I expect you have just as many (if not more) ~~distractions~~ _commitments_ competing for your attention as I do, but I'm still happy to propose an implementation for your review when I can find some time.

**Comment by leycec** [comment]
I'm actually a big fan of shame! Shame's great. Love shame. More shame? Yes, please. Japan's the safest, healthiest, and funn(i)est Westphalian nation state I've ever been graced to live in – and it's all predicate on *informal shame* rather than *formal punishment.*

The core idea here is that society can (almost entirely) avoid formal punishment by just doubling down on informal shame instead. When shame carries profound (but temporary) real-world consequences, formal punishment and all its attendant miseries becomes but a perfunctory footnote you hope to never exercise.

This is *kinda* the hidden premise of the [Greatest Film of All Time That You Must Absolutely Watch but Never with Anyone You Hope to Speak to Again](https://www.imdb.com/title/tt0056058/?ref_=fn_al_tt_1). Bro, can you feel this palpable tension?

![this is tense](https://s3.amazonaws.com/static.rogerebert.com/uploads/review/primary_image/reviews/great-movie-harakiri-1962/EB20120223REVIEWS08120229987AR.jpg)
<sup>*tense situation here*</sup>

![this also tense](https://i0.wp.com/projectedfigures.com/wp-content/uploads/2015/03/vlcsnap-2011-07-04-12h21m54s65-2.png?zoom=2&resize=591%2C267)
<sup>*tension intensifies*</sup>

> ...I'm still happy to propose an implementation for your review when I can find some time.

This brings me manifold happiness beyond any you could imagine.
A PR doing everything for me that I should have already done would be the bee's arthritic knees and just *fabulous.* :magic_wand: 

Oh, and fully unrelatedly – I'd love to highlight `numerary` in our FAQ! I keep meaning to submit an issue to your issue tracker... but then life intensifies. When I *finally* do, would you mind helping me hash out a FAQ entry on "How to type-check numbers?" The rabbit hole goes deep here and you're the resident expert. I defer to your magnanimous wisdom.

**Reply by posita** [reply]
You do you. I'll help in whatever way I can.

I have put your film recommendation to the proverbial "top of my queue". Now I only have to find time to watch a screen that is not connected to my keyboard, and we'll be good. By the way, I mistook …

> … Greatest Film of All Time That You Must Absolutely Watch but Never with Anyone You Hope to Speak to Again …

… as describing (in part) [_Requiem for a Dream_](https://www.imdb.com/title/tt0180093/), but that may hold a similar-sounding, but significantly different title, which is "Best Film that I Am Unable to Bring Myself to Ever Watch Again".

> Oh, and fully unrelatedly – I'd love to highlight `numerary` in our FAQ! I keep meaning to submit an issue to your issue tracker... but then life intensifies. When I _finally_ do, would you mind helping me hash out a FAQ entry on "How to type-check numbers?" The rabbit hole goes deep here and you're the resident expert. I defer to your magnanimous wisdom.

I'm happy to write an entry, but maybe we should suss out posita/numerary#7 _et al._, which I am very much onboard with.

**Comment by leycec** [comment]
> *Requiem for a Dream*

Pretty sure humanity collectively agreed to pretend that never happened. The closing five minutes is permanently seared into the thin jelly of my eyeballs – along with the closing five minutes to **every other Darren Aronofsky film ever.**

Pretty sure humanity collectively agreed to pretend those never happened, too.

![ohshit](https://images.mubicdn.net/images/film/1647/cache-8967-1546736427/image-w1280.jpg?size=960x)
<sup>pretty sure this never happened... *but not certain*</sup>
